### DISCUSSION TITLE
Berry Script: No return for light.get() ?

### DISCUSSION THREAD

**Discussion by ozett** [arendst/Tasmota#17878] [Q&A] [answered]
### PROBLEM DESCRIPTION
i started to use berry script and expect to get a return for `light.get()`.
maybe my misreading of the documentation - or a bug?
maybe berry´s light is not applicable for a LED-configred GPIO?

docu: 
<kbd>![image](https://user-images.githubusercontent.com/5937811/216761174-f5a107ab-6a33-4153-96cf-38ed843b1fd8.png)

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ x] Device used (e.g., Sonoff Basic): lolin lite
- [ ] Tasmota binary firmware version number used: 12.3.1
  - [x ] Pre-compiled
  - [ ] Self-compiled
- [ x] Flashing tools used: webinstaller
- [x ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
10:57:30.785 CMD: Backlog Template; Module; GPIO 255
10:57:30.812 MQT: stat/tasmota_3DF6D4/RESULT = {"NAME":"ESP32-DevKit","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}
10:57:31.039 MQT: stat/tasmota_3DF6D4/RESULT = {"Module":{"1":"ESP32-DevKit"}}
10:57:31.294 MQT: stat/tasmota_3DF6D4/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO6":{"0":"None"},"GPIO7":{"0":"None"},"GPIO8":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO11":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"288":"Led1"},"GPIO17":{"0":"None"},"GPIO18":{"0":"None"},"GPIO19":{"0":"None"},"GPIO20":{"0":"None"},"GPIO21":{"0":"None"},"GPIO22":{"0":"None"},"GPIO23":{"0":"None"},"GPIO24":{"0":"None"},"GPIO25":{"0":"None"},"GPIO26":{"0":"None"},"GPIO27":{"0":"None"},"GPIO32":{"0":"None"},"GPIO33":{"0":"None"},"GPIO34":{"0":"None"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO37":{"0":"None"},"GPIO38":{"0":"None"},"GPIO39":{"0":"None"}}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
10:58:18.202 CMD: Backlog Rule1; Rule2; Rule3
10:58:18.232 MQT: stat/tasmota_3DF6D4/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
10:58:18.458 MQT: stat/tasmota_3DF6D4/RESULT = {"Rule2":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
10:58:18.659 MQT: stat/tasmota_3DF6D4/RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}

```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
10:58:39.520 CMD: Status 0
10:58:39.530 MQT: stat/tasmota_3DF6D4/STATUS = {"Status":{"Module":1,"DeviceName":"lolin-lite-flashled","FriendlyName":["lolin-lite-flashled"],"Topic":"tasmota_3DF6D4","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
10:58:39.540 MQT: stat/tasmota_3DF6D4/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"https://ota.tasmota.com/tasmota32/release/tasmota32.bin","RestartReason":"Software reset CPU","Uptime":"0T00:19:42","StartupUTC":"2023-02-04T09:38:57","Sleep":50,"CfgHolder":4617,"BootCount":6,"BCResetTime":"2023-02-03T20:01:13","SaveCount":32}}
10:58:39.546 MQT: stat/tasmota_3DF6D4/STATUS2 = {"StatusFWR":{"Version":"12.3.1(tasmota)","BuildDateTime":"2023-01-24T09:38:00","Core":"2_0_5_3","SDK":"v4.4.3","CpuFrequency":160,"Hardware":"ESP32-D0WDQ6-V3 rev.3","CR":"411/699"}}
10:58:39.551 MQT: stat/tasmota_3DF6D4/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["fo46io",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
10:58:39.569 MQT: stat/tasmota_3DF6D4/STATUS4 = {"StatusMEM":{"ProgramSize":1479,"Free":1400,"Heap":138,"StackLowMark":3,"PsrMax":0,"PsrFree":0,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":"DIO","Features":["00000809","9F9AD7CF","0015A001","B7F7BFCF","05DA9BC0","E0360DC7","400840D2","20200000","D43C482D","000000A1"],"Drivers":"1,2,3,4,5,7,8,9,10,11,12,14,16,17,20,21,24,26,27,29,34,35,38,50,52,59,60,62,63,82,86,87,88","Sensors":"1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,28,31,34,37,39,40,42,43,45,51,52,55,56,58,59,64,66,67,74,85,92,95,98,127"}}
10:58:39.582 MQT: stat/tasmota_3DF6D4/STATUS5 = {"StatusNET":{"Hostname":"tasmota-3DF6D4-5844","IPAddress":"192.168.14.15","Gateway":"192.168.14.9","Subnetmask":"255.255.255.0","DNSServer1":"192.168.14.9","DNSServer2":"8.8.8.8","Mac":"B4:8A:0A:3D:F6:D4","Ethernet":{"Hostname":"","IPAddress":"0.0.0.0","Gateway":"0.0.0.0","Subnetmask":"0.0.0.0","DNSServer1":"0.0.0.0","DNSServer2":"0.0.0.0","Mac":"00:00:00:00:00:00"},"Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
10:58:39.593 MQT: stat/tasmota_3DF6D4/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.14.204","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_3DF6D4","MqttUser":"DVES_USER","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
10:58:39.606 MQT: stat/tasmota_3DF6D4/STATUS7 = {"StatusTIM":{"UTC":"2023-02-04T09:58:39","Local":"2023-02-04T10:58:39","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"08:16","Sunset":"17:51"}}
10:58:39.619 MQT: stat/tasmota_3DF6D4/STATUS10 = {"StatusSNS":{"Time":"2023-02-04T10:58:39"}}
10:58:39.632 MQT: stat/tasmota_3DF6D4/STATUS11 = {"StatusSTS":{"Time":"2023-02-04T10:58:39","Uptime":"0T00:19:42","UptimeSec":1182,"Heap":142,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":44},"Wifi":{"AP":1,"SSId":"fo46io","BSSId":"F4:92:BF:A1:06:DD","Channel":1,"Mode":"11n","RSSI":78,"Signal":-61,"LinkCount":1,"Downtime":"0T00:00:04"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
10:58:39.520 CMD: Status 0
10:58:39.530 MQT: stat/tasmota_3DF6D4/STATUS = {"Status":{"Module":1,"DeviceName":"lolin-lite-flashled","FriendlyName":["lolin-lite-flashled"],"Topic":"tasmota_3DF6D4","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
10:58:39.540 MQT: stat/tasmota_3DF6D4/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"https://ota.tasmota.com/tasmota32/release/tasmota32.bin","RestartReason":"Software reset CPU","Uptime":"0T00:19:42","StartupUTC":"2023-02-04T09:38:57","Sleep":50,"CfgHolder":4617,"BootCount":6,"BCResetTime":"2023-02-03T20:01:13","SaveCount":32}}
10:58:39.546 MQT: stat/tasmota_3DF6D4/STATUS2 = {"StatusFWR":{"Version":"12.3.1(tasmota)","BuildDateTime":"2023-01-24T09:38:00","Core":"2_0_5_3","SDK":"v4.4.3","CpuFrequency":160,"Hardware":"ESP32-D0WDQ6-V3 rev.3","CR":"411/699"}}
10:58:39.551 MQT: stat/tasmota_3DF6D4/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["fo46io",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
10:58:39.569 MQT: stat/tasmota_3DF6D4/STATUS4 = {"StatusMEM":{"ProgramSize":1479,"Free":1400,"Heap":138,"StackLowMark":3,"PsrMax":0,"PsrFree":0,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":"DIO","Features":["00000809","9F9AD7CF","0015A001","B7F7BFCF","05DA9BC0","E0360DC7","400840D2","20200000","D43C482D","000000A1"],"Drivers":"1,2,3,4,5,7,8,9,10,11,12,14,16,17,20,21,24,26,27,29,34,35,38,50,52,59,60,62,63,82,86,87,88","Sensors":"1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,28,31,34,37,39,40,42,43,45,51,52,55,56,58,59,64,66,67,74,85,92,95,98,127"}}
10:58:39.582 MQT: stat/tasmota_3DF6D4/STATUS5 = {"StatusNET":{"Hostname":"tasmota-3DF6D4-5844","IPAddress":"192.168.14.15","Gateway":"192.168.14.9","Subnetmask":"255.255.255.0","DNSServer1":"192.168.14.9","DNSServer2":"8.8.8.8","Mac":"B4:8A:0A:3D:F6:D4","Ethernet":{"Hostname":"","IPAddress":"0.0.0.0","Gateway":"0.0.0.0","Subnetmask":"0.0.0.0","DNSServer1":"0.0.0.0","DNSServer2":"0.0.0.0","Mac":"00:00:00:00:00:00"},"Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
10:58:39.593 MQT: stat/tasmota_3DF6D4/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.14.204","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_3DF6D4","MqttUser":"DVES_USER","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
10:58:39.606 MQT: stat/tasmota_3DF6D4/STATUS7 = {"StatusTIM":{"UTC":"2023-02-04T09:58:39","Local":"2023-02-04T10:58:39","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"08:16","Sunset":"17:51"}}
10:58:39.619 MQT: stat/tasmota_3DF6D4/STATUS10 = {"StatusSNS":{"Time":"2023-02-04T10:58:39"}}
10:58:39.632 MQT: stat/tasmota_3DF6D4/STATUS11 = {"StatusSTS":{"Time":"2023-02-04T10:58:39","Uptime":"0T00:19:42","UptimeSec":1182,"Heap":142,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":44},"Wifi":{"AP":1,"SSId":"fo46io","BSSId":"F4:92:BF:A1:06:DD","Channel":1,"Mode":"11n","RSSI":78,"Signal":-61,"LinkCount":1,"Downtime":"0T00:00:04"}}}

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
type in berry console: `light.get()`

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
a return value like its happening with the power on berry-console
tasmota.get_power()
-> https://tasmota.github.io/docs/Berry/#lights-and-relays

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
![image](https://user-images.githubusercontent.com/5937811/216760889-b5ec8bfb-f0b4-496b-937e-2ecc4f23bbf5.png)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment, marked as answer]
A led is not a light

Lights are:
- relay
- pwm
- ws2812

Led is just an indicator

**Reply by sfromis** [reply]
Strictly speaking, a `Relay` pin does not count as a light in Tasmota, but various controllers commonly seen in "smart" bulbs do.

**Comment by ozett** [comment]
Yeah, just tried to switch to PWM and i got a return.

do you think thats a common understanting that `LED` is **not** a `LIGHT` ?

anyways . big thanks for the lightningspeed response. will dig further into berry... 😄 

<kbd>![image](https://user-images.githubusercontent.com/5937811/216761608-22cbcfde-4d5c-404e-be30-93eb76e81383.png)

**Reply by sfromis** [reply]
To me, the common understanding of a "LED" (singular noun, light emitting diode) is a small thing to indicate status, not a lamp. You may have a LED-based bulb, but that does not mean that I'd call the bulb a "LED".

OTOH, if you configure a PWM, Tasmota does know if that happens to be a singular LED, while the more common setup is a strip, lamp, or similar.

**Reply by ozett** [reply]
as i am not a native english speaker, i started to assume that a (LED) **Light**-**E**mitting-**D**iode can somehow be controlled with scripting code that uses the keyword "**Light**".  also the fine semantic differentiation between light, lamp and bulb did not help me.

maybe it would be easear for low-coder like me to have also an alias for 
* PWM.get/set()
* relay,get/set()
* ws2812).get/set()

in the scripting context of berry? 
so to say, commands straightly named after the gpio-drivername?

i did read the berry-documentation https://tasmota.github.io/docs/Lights/ , but did not find anything clear to me that explains how to configure the module's gpios for the use of a corersponding berry command. as for the "light" getter/setter thats now clearer to me.

**Reply by barbudor** [reply]
I'm French, sfromis is Dutch, many Germans and others around. English is, for the best or the worse, the Esperanto of technology.

But if you start with Tasmota by jumping directly into Berry without first getting a hint on the Tasmota base concepts, you'll for sure have other surprises.
Berry comes on top of Tasmota. It's not Arduino nor Micropython.

**Reply by sfromis** [reply]
The starting point should indeed be to configure Tasmota with a working light (which is not a "Relay"). After that, you can go on to work on how to use it with the Berry language, where you do not need to care about whether it is a PWM, bulb or WS2812 (unless you want addressable led support).

**Reply by barbudor** [reply]
Nevertheless, the 1st category of lights in the [doc](https://tasmota.github.io/docs/Lights/#light-types) is

![image](https://user-images.githubusercontent.com/2996042/216776909-58bddb5d-44df-4779-b135-50ce39470001.png)

:roll_eyes:

**Reply by sfromis** [reply]
While you of course can hook a lamp up to a relay, this does mean that the Tasmota code sees it as part of the lights feature, and the `lt_st` field in the discovery config payload is zero for "no light subtype". Thus, the Berry feature for interacting with a light (lamp) does not apply. Even when using `SetOption30` to "Enforce Home Assistant auto-discovery as light", I'm still seeing `"lt_st":0` and `light.get()` returning `nil`. 

Arguably, this might be a bug, since the old `xdrv_12_home_assistant.ino` does use `SetOption30` to allow the user to treat a relay as a light, which seems to mean announcing it as "light" instead of "switch". Looks like a fix is relevant, either to docs or code, but having the relay announced as a light would still not activate the Tasmota lights feature. As things are now, I can't see `SetOption30` as making any difference, unless doing a custom build for old HA MQTT discovery.

**Reply by ozett** [reply]
experimenting more with berry, i have PWM configured LED on pin16.
now i want to switch it on/off and wondering wich berry-cmd will work. 
Not an easy task, switching a LED.

trying to use gpio-module for this:
<kbd> ![image](https://user-images.githubusercontent.com/5937811/216813848-6b35ed9a-b9a0-4fa3-bd21-7d01083afe83.png)

Now trying as PWM. as PWM is not a light, light.get() should not work? and it should not work to swich the led on/off?
but instead it first throws an error, but than toggles the LED and returns the power-state.

should it work this way, like i expected at the begining, that the light.set() can be used to switch the LED?
or still under development?

would love to see this to work the same way: `gpio.digital_write(gpio.pin(gpio.PWM1), gpio.HIGH)` [from docu](https://tasmota.github.io/docs/Berry/#gpio-module)

<kbd>![image](https://user-images.githubusercontent.com/5937811/216814696-be3368dd-0199-4527-955e-61887d853078.png)

is there anything wrong with my undertakings?

<kbd>![berry-ledget](https://user-images.githubusercontent.com/5937811/216814371-39529d62-e847-4f25-9449-327fca75aa01.gif)

and i found this also to switch the PWM1=LED on pin 16:

```
//PWMS logical GPiOs, so cmd (PWM1, relay, ...)
gpio.set_pwm(16,0)
gpio.set_pwm(16,1023)
tasmota.set_power(0, false)
tasmota.set_power(0, true)
```
(and a little confusing to me is the syntax, that `tasmota.set_power()` uses **not** the pysical gpio-numbering, but instead a `list of bool`of the configured GPIOs ([docu](https://tasmota.github.io/docs/Berry/#functions-to-manage-relays-and-lights)) . but maybe thats BERRY 😄

**Reply by sfromis** [reply]
Indeed, it does not make sense to use `gpio.digital_write` on a PWM pin, unless you can actually produce PWM signals, meaning hundreds of writes a second with closely managed timing.

When it comes to `tasmota.set_power`, this is not addressing a gpio pin at all. Instead it is a power channel, which can be many different things, including I2C displays or led controller chips (used in many bulbs), where there just is no specific pin for power on/off. Thus a relay, PWM or WS2812 is just one of the possible meanings. As this is the same as when using the `Power` command outside of Berry, this is not a quirk of this language implementetion.

When it comes to `gpio.set_pwm`, that is meant for using Berry to control a PWM **instead of** it acting as a "light" in Tasmota, meaning only when configuring the gpio with PWM6 or higher, as the first 5 are used to form a light. Thus things configured on PWM1-5 is very much a light, unless also having things like WS2812 configured at the same time.

You did not tell what gpio config you were using while having the errors in that reading-unfriendly gif animation, when getting:
`BRY: Exception> 'type_error' - 'nil' value is not callable`
 Could be a bug in the `light` module of Tasmota. 

Again, you should really start making the light work **outside of** Berry, to understand how it acts, before then working to control it with the Berry language.

More generally, when you config with PWM1-5 or WS2812-1, this will be seen as a light in Tasmota, where using the `light` module in Berry applies. With Berry, you can using other configs (PWM6 and above, and WS2812 index 2 and above) to manage light outputs outside of what can be done with "classic" Tasmota, using specific Berry features.

When managing PWM channels as a light, the number of PWM pins (from PWM1-5) in use will determine whether the light is white, CCT, RGB, RGBW or RGBCCT, instead of each PWM channel used separately. What you get with Berry and PWM is the option to control additional PWM channels outside of this, but without the features to manage multiple channels as a single light.

**Reply by ozett** [reply]
thanks for all that background information, really appreciate it.

>Indeed, it does not make sense to use gpio.digital_write on a PWM pin, unless you can actually produce PWM signals,

but for my perspective i only wanted to write '0' and '1023'  on PWM, keep all pwm-possibilites and only mimic on/off. but i see your brighter picture.

>You did not tell what gpio config you were using while having the errors

you mean: `Backlog Template; Module; GPIO 255` this? ->  ' "GPIO16":{"416":"PWM1"}'  -> full ->
```
12:59:04.418 CMD: Backlog Template; Module; GPIO 255
12:59:04.420 SRC: WebConsole from 192.168.14.135
12:59:04.422 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 26, Pld -99, Data 'Template; Module; GPIO 255'
12:59:04.446 SRC: Backlog
12:59:04.448 CMD: Grp 0, Cmd 'TEMPLATE', Idx 1, Len 0, Pld -99, Data ''
12:59:04.457 MQT: stat/tasmota_3DF6D4/RESULT = {"NAME":"ESP32-DevKit","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}
12:59:04.651 SRC: Backlog
12:59:04.654 CMD: Grp 0, Cmd 'MODULE', Idx 1, Len 0, Pld -99, Data ''
12:59:04.662 MQT: stat/tasmota_3DF6D4/RESULT = {"Module":{"1":"ESP32-DevKit"}}
12:59:04.858 SRC: Backlog
12:59:04.860 CMD: Grp 0, Cmd 'GPIO', Idx 1, Len 3, Pld 255, Data '255'
12:59:04.868 MQT: stat/tasmota_3DF6D4/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO6":{"0":"None"},"GPIO7":{"0":"None"},"GPIO8":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO11":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"416":"PWM1"},"GPIO17":{"0":"None"},"GPIO18":{"0":"None"},"GPIO19":{"0":"None"},"GPIO20":{"0":"None"},"GPIO21":{"0":"None"},"GPIO22":{"0":"None"},"GPIO23":{"0":"None"},"GPIO24":{"0":"None"},"GPIO25":{"0":"None"},"GPIO26":{"0":"None"},"GPIO27":{"0":"None"},"GPIO32":{"0":"None"},"GPIO33":{"0":"None"},"GPIO34":{"0":"None"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO37":{"0":"None"},"GPIO38":{"0":"None"},"GPIO39":{"0":"None"}}
12:59:08.731 WIF: Checking connection...
```
please advise me for debugging information, if you like

> More generally, when you config with PWM1-5 or WS2812-1, this will be seen as a light in Tasmota, where using the light module in Berry applies.

i read this 2 times, didnt got it, why PWM 1-5 are "Lights", but now i think this is great information. 
hard to find in the docu and hard to see withour your information, that berry lights.module is usable on pwm1-5 only
https://tasmota.github.io/docs/Lights/#pwm6 and the channels for 1-5 here:
https://tasmota.github.io/docs/Lights/#1-channel-dimmable-light

> Again, you should really start making the light work outside of Berry, to understand how it acts, before then working to control it with the Berry language

i still think of tasmota as a everyones darling (for the home-automation), and not only for the most experienced coder. so there are simpler ways to use tasmota, but berry seems powerful and a modern way. why not fiddle with this? is it meant only for the most experienced people? if not, i see thats hard to cope with the newbies.., sorry for that. 

my starting point is to switch three times a led on and off, each on and off for nearly 50ms.
as commands and backlog takes longer i only found scripting a solution to that short times blinker.

since so many people work with tasmota, i did not found tricks for that on the issues and discussions, so i thought i would try the most modern part of tasmota and start with berry. alongside it's great fun and i see other possible uses with berry for my other tasmotas. 

but maybe there is still a hidden but totally simple way to make such a led flash at an interval of nearely or less than 50 ms?

**Reply by sfromis** [reply]
While Berry is powerful (and adds a lot of fancy things you can do), it sits on top of "classic" Tasmota, and does not try to create a different model for regular features. Therefore, it does not escape how PWM1-5 are "captured" as a light, before Berry even is part of the picture. Hence it is easier to grok how Berry works with Tasmota when you already know about the directly relevant features of Tasmota, in this case lights. While you do not need to understand "all about lights", you did now find the part of the docs with the story about PWM1-5.

And with your config being with one PWM channel, on PWM1, this means that Tasmota configures this as a simple "white" light, which you can dim up and down using the regular light feature, and this should work with the `light` module of Tasmota. The error message you got when doing reasonable things to try that does not look like you "doing it wrong".

You using `gpio.set_pwm` was quite reasonable, except for using PWM1 which was "taken up" by been seen as a light.

**Reply by ozett** [reply]
yeah, thanks for `lighting` this all up to me with valuable information.

still wondering if there is a need to debug my error on `light.get/set` ?  can I fix the error myself somehow?

i also see in some places that you need an basic understanding of c/c++. although the first steps with tasmota always seem so easy.
But if you don't give up, you will experience (even with Berry) the great revelation of the adjustments of simple functions. but all this is just fun, thanks to the wonderful help here in the forum. i still haven't lost the desire to keep experimenting with berry. First of all, thank you very much. I hope others will benefit from the discussion as I have here.

**Comment by barbudor** [comment]
I believe the documentation provides quite some hints in what is Light for Tasmota
https://tasmota.github.io/docs/Lights/

Glad that solve your problem

For help request and general how-to, please start a Discussion
Issues should focus on qualified bugs
Thanks for your understanding
